File: sdk/typescript/src/guard/topic-restriction.ts
/**
 * Topic restriction guard.
 *
 * Enforces allow/deny topic lists using a user-provided topic classifier.
 *
 * Must stay in sync with: sdk/python/src/openlit/guard/topic_restriction.py
 */

import { Guard, GuardConfigError, GuardPhase, GuardResult, makeGuardResult, GuardOptions } from './base';

export interface TopicRestrictionOptions extends GuardOptions {
  classifier: (text: string) => string;
  allowed?: string[];
  denied?: string[];
}

export class TopicRestriction extends Guard {
  readonly name = 'topic_restriction';
  readonly phases = [GuardPhase.PREFLIGHT];

  private readonly _classifier: (text: string) => string;
  private readonly _allowed: Set<string> | null;
  private readonly _denied: Set<string> | null;

  constructor(opts: TopicRestrictionOptions) {
    super({ action: opts.action ?? 'deny', maxScanLength: opts.maxScanLength });

    if (typeof opts.classifier !== 'function') {
      throw new GuardConfigError("TopicRestriction requires a callable 'classifier'");
    }
    if (opts.allowed && opts.denied) {
      throw new GuardConfigError("Provide either 'allowed' or 'denied', not both");
    }
    if (!opts.allowed && !opts.denied) {
      throw new GuardConfigError("Provide either 'allowed' or 'denied' topic list");
    }

    this._classifier = opts.classifier;
    this._allowed = opts.allowed ? new Set(opts.allowed.map((t) => t.toLowerCase())) : null;
    this._denied = opts.denied ? new Set(opts.denied.map((t) => t.toLowerCase())) : null;
  }

  evaluate(text: string): GuardResult {
    const topic = this._classifier(text).toLowerCase().trim();

    if (this._allowed !== null && !this._allowed.has(topic)) {
      return makeGuardResult({
        action: this._action,
        score: 1.0,
        guardName: this.name,
        classification: topic,
        explanation: `Topic '${topic}' is not in the allowed list`,
      });
    }

    if (this._denied !== null && this._denied.has(topic)) {
      return makeGuardResult({
        action: this._action,
        score: 1.0,
        guardName: this.name,
        classification: topic,
        explanation: `Topic '${topic}' is in the denied list`,
      });
    }

    return makeGuardResult({ guardName: this.name, classification: topic });
  }
}

File: sdk/python/src/openlit/guard/topic_restriction.py
"""
Topic restriction guard.

Enforces allow/deny topic lists using a user-provided topic classifier.
Phases: preflight only.
"""

from __future__ import annotations

from typing import Callable, List, Optional, Set

from openlit.guard._base import (
    Guard,
    GuardConfigError,
    GuardPhase,
    GuardResult,
)


class TopicRestriction(Guard):
    """
    Blocks prompts that don't match allowed topics or that match denied topics.

    Parameters
    ----------
    classifier : callable
        ``Callable[[str], str]`` that returns a topic label for the given text.
    allowed : list[str], optional
        Only these topics are permitted.  Mutually exclusive with ``denied``.
    denied : list[str], optional
        These topics are blocked; everything else is allowed.
    action : str
        ``"deny"`` (default) or ``"warn"``.
    """

    name = "topic_restriction"
    phases = (GuardPhase.PREFLIGHT,)

    def __init__(
        self,
        classifier: Callable[[str], str],
        allowed: Optional[List[str]] = None,
        denied: Optional[List[str]] = None,
        action: str = "deny",
        **kwargs,
    ):
        super().__init__(action=action, **kwargs)
        if not callable(classifier):
            raise GuardConfigError("TopicRestriction requires a callable 'classifier'")
        if allowed and denied:
            raise GuardConfigError("Provide either 'allowed' or 'denied', not both")
        if not allowed and not denied:
            raise GuardConfigError("Provide either 'allowed' or 'denied' topic list")

        self._classifier = classifier
        self._allowed: Optional[Set[str]] = (
            {t.lower() for t in allowed} if allowed else None
        )
        self._denied: Optional[Set[str]] = (
            {t.lower() for t in denied} if denied else None
        )

    def evaluate(self, text: str) -> GuardResult:
        topic = self._classifier(text).lower().strip()

        if self._allowed is not None and topic not in self._allowed:
            return GuardResult(
                action=self._action,
                score=1.0,
                guard_name=self.name,
                classification=topic,
                explanation=f"Topic '{topic}' is not in the allowed list",
            )

        if self._denied is not None and topic in self._denied:
            return GuardResult(
                action=self._action,
                score=1.0,
                guard_name=self.name,
                classification=topic,
                explanation=f"Topic '{topic}' is in the denied list",
            )

        return GuardResult(guard_name=self.name, classification=topic)
